crates/wasm-mutate/src/mutators/codemotion/ir/mod.rs

```diff
@@ -0,0 +1,501 @@
+    Block(Vec<usize>, TypeOrFuncType),
+    /// Special node to wrap the root nodes of the AST
+    Root(Vec<usize>),
+}
+
+impl ASTBuilder {
+    /// Returns an AST from the operators collected from the Wasm function
+    pub fn build_ast<'a>(&self, operators: &'a [OperatorAndByteOffset]) -> crate::Result<AST> {
+        let (root, nodes, ifs, loops, blocks) = self.parse(operators)?;
+        Ok(AST::new(root, nodes, ifs, loops, blocks))
+    }
+
+    /// Parsing algorith to construct the AST
+    fn parse<'a>(
+        &self,
+        operators: &'a [OperatorAndByteOffset],
+    ) -> crate::Result<(usize, Vec<Node>, Vec<usize>, Vec<usize>, Vec<usize>)> {
+        #[derive(Debug)]
+        enum State {
+            If,
+            Else,
+            Loop,
+            Block,
+            Root,
+        }
+
+        let nodes = Vec::new();
+
+        let mut ifs = Vec::new();
+        let mut loops = Vec::new();
+        let mut blocks = Vec::new();
+
+        let nodes = RefCell::new(nodes);
+
+        let mut frame_stack = vec![(State::Root, None)];
+        let state_stack = RefCell::new(Vec::new());
+        let current_parsing = RefCell::new(Vec::new());
+
+        let push_state = || {
+            state_stack.borrow_mut().push(current_parsing.clone());
+            *current_parsing.borrow_mut() = vec![]
+        };
+        let pop_state = || {
+            state_stack
+                .borrow_mut()
+                .pop()
+                .expect("Missing state")
+                .borrow()
+                .clone()
+        };
+
+        let mut current_code = Range::new(0, 0);
+
+        for (idx, (operator, _)) in operators.iter().enumerate() {
+            match operator {
+                Operator::If { ty } => {
+                    // push current code
+                    if current_code.end != current_code.start {
+                        let id = nodes.borrow().len();
+                        nodes.borrow_mut().push(Node::Code {
+                            range: current_code,
+                        });
+                        current_parsing.borrow_mut().push(id);
+                    }
+                    current_code.start = idx + 1;
+                    current_code.end = idx + 1;
+                    push_state();
+                    frame_stack.push((State::If, Some(ty)));
+                }
+                Operator::Else => {
+                    if current_code.end != current_code.start {
+                        let id = nodes.borrow().len();
+                        nodes.borrow_mut().push(Node::Code {
+                            range: current_code,
+                        });
+                        current_parsing.borrow_mut().push(id);
+                    }
+                    current_code.start = idx + 1;
+                    current_code.end = idx + 1;
+                    push_state();
+                    frame_stack.push((State::Else, None));
```

Comment:
It seems like `push_state` should take `State` and `Option<TypeOrFuncType>` arguments, since every `push_state()` call has a corresponding `frame_stack.push(..)` call.

Adding them as arguments to `push_state` helps make sure that the stacks can't get out of sync.